filebeat/input/state.go

```diff
@@ -20,3 +26,76 @@ func NewFileState(fileInfo os.FileInfo, path string) FileState {
 		FileStateOS: GetOSFileState(fileInfo),
 	}
 }
+
+// States handles list of FileState
+type States struct {
+	states []FileState
+	mutex  sync.Mutex
+}
+
+func NewStates() *States {
+	return &States{
+		states: []FileState{},
+	}
+}
+
+// Update updates a state. If previous state didn't exist, new one is created
+func (s *States) Update(newState FileState) {
+
+	index, oldState := s.FindPrevious(newState)
+
+	// Only block after FindPrevious as FindPrevious has its own lock
```

Comment:
Do you mean if it could happen that after FindPrevious an other method gets the lock and updates the state? Potentially yes I would say, but not 100% sure about the affect. I think what would happen is that a newer state could be overwritten by an older one. This would have an affect until a new state comes in which then brings it up-to-date again. So in between it could happen, that an older state exists.